### PR TITLE
Add new stats endpoint

### PR DESCRIPTION
**Pull request by tianj7**
https://ctds-planx.atlassian.net/browse/VADC-1576

### New Features
Add new stats endpoint for cohort and concepts

### CODE REVIEW

File: utils/parsing.go [hunk 1]
```diff
@@ -247,12 +266,12 @@ func ParseSourceAndCohortId(c *gin.Context) (int, int, error) {
...
-func GetConceptIdsAndCohortPairsAsSeparateLists(conceptIdsAndCohortPairs []interface{}) ([]int64, []CustomDichotomousVariableDef) {
-	conceptIds := []int64{}
+func GetConceptIdsAndCohortPairsAsSeparateLists(conceptIdsAndCohortPairs []interface{}) ([]CustomConceptVariableDef, []CustomDichotomousVariableDef) {
+	conceptIds := []CustomConceptVariableDef{}
 	cohortPairs := []CustomDichotomousVariableDef{}
 	for _, item := range conceptIdsAndCohortPairs {
 		switch convertedItem := item.(type) {
-		case int64:
+		case CustomConceptVariableDef:
 			conceptIds = append(conceptIds, convertedItem)
```
Comment (pieterlukasse) [inline utils/parsing.go:275]: I think it is better to rename `conceptIds` here, as it is not a list of ids anymore, but a list of CustomConceptVariableDef

File: utils/parsing.go [hunk 2]
```diff
@@ -267,7 +286,8 @@ func ParseSourceIdAndCohortIdAndVariablesList(c *gin.Context) (int, int, []int64
 	if err != nil {
 		return -1, -1, nil, nil, err
 	}
-	conceptIds, cohortPairs := GetConceptIdsAndCohortPairsAsSeparateLists(conceptIdsAndCohortPairs)
+	conceptIdsAndValues, cohortPairs := GetConceptIdsAndCohortPairsAsSeparateLists(conceptIdsAndCohortPairs)
+	conceptIds := ExtractConceptIdsFromCustomConceptVariablesDef(conceptIdsAndValues)
```
Comment (pieterlukasse) [inline utils/parsing.go:290]: maybe move `ExtractConceptIdsFromCustomConceptVariablesDef` into `GetConceptIdsAndCohortPairsAsSeparateLists` ?

File: models/helper.go [hunk 3]
```diff
@@ -25,6 +25,25 @@ func QueryFilterByConceptIdsHelper(query *gorm.DB, sourceId int, filterConceptId
...
+func QueryFilterByConceptIdsAndValuesHelper(query *gorm.DB, sourceId int, filterConceptIdsAndValues []utils.CustomConceptVariableDef,
+	omopDataSource *utils.DbAndSchema, resultSchemaName string, personIdFieldForObservationJoin string) *gorm.DB {
+	// iterate over the filterConceptIds, adding a new INNER JOIN and filters for each, so that the resulting set is the
+	// set of persons that have a non-null value for each and every one of the concepts:
+	for i, filterConceptIdAndValue := range filterConceptIdsAndValues {
+		observationTableAlias := fmt.Sprintf("observation_filter_%d", i)
+		log.Printf("Adding extra INNER JOIN with alias %s", observationTableAlias)
+		query = query.Joins("INNER JOIN "+omopDataSource.Schema+".observation_continuous as "+observationTableAlias+omopDataSource.GetViewDirective()+" ON "+observationTableAlias+".person_id = "+personIdFieldForObservationJoin).
+			Where(observationTableAlias+".observation_concept_id = ?", filterConceptIdAndValue.ConceptId).
+			Where(GetConceptValueNotNullCheckBasedOnConceptType(observationTableAlias, sourceId, filterConceptIdAndValue.ConceptId))
```
Comment (pieterlukasse) [inline models/helper.go:38]: you could move this to else after the `if len(filterConceptIdAndValue.ConceptValues) > 0` below

**Review by pieterlukasse — COMMENTED** [review]
thanks @tianj7 . I added some minor improvement suggestions

File: models/helper.go (hunk 3 above)
Comment (tianj7) [reply models/helper.go:38]: Hmm I don't think this code belongs to an else block after the check. The reason is we still need to limit the observation to this particular conceptid. If the value filters are passed in we then AND the clause for filter by value. It's not an either/or relationship.

If the value_as_concept_id is shared with other concepts we could potentially pick up with unwanted rows relating other concept ids.

File: utils/parsing.go (hunk 1 above)
Comment (tianj7) [reply utils/parsing.go:275]: Updated

File: models/helper.go (hunk 3 above)
Comment (pieterlukasse) [reply models/helper.go:38]: I meant: no need to check for "not null" if we are already checking for a specific value.

File: utils/parsing.go (hunk 2 above)
Comment (pieterlukasse) [reply utils/parsing.go:290]: but you can still use it inside `GetConceptIdsAndCohortPairsAsSeparateLists`, right?

Comment (tianj7) [reply utils/parsing.go:290]: `ExtractConceptIdsFromCustomConceptVariablesDef` is created to give a convenient way to plug things into existing functions which only support concept ids to limit the scope of this PR.

`GetConceptIdsAndCohortPairsAsSeparateLists` is used in `ParseConceptIdsAndDichotomousDefs` which in turn is used in the new stat endpoint, so we need them to pass back both ids and values. If we use 'ExtractConceptIdsFromCustomConceptVariablesDef inside `GetConceptIdsAndCohortPairsAsSeparateLists` then we will loss the value part.

File: models/helper.go (hunk 3 above)
Comment (tianj7) [reply models/helper.go:38]: yes I see what you mean now, updated

**Review by pieterlukasse — APPROVED** [review]

**Pull request merged**